nemo/collections/nlp/models/language_modeling/megatron_base_model.py

```diff
@@ -805,19 +816,18 @@ def build_model_parallel_config(self) -> ModelParallelConfig:
 
         return model_parallel_config
 
-    def _prefetch(self, iterator):
-        """Checks if the iterator still has elements to return.
-        Used in models using dataloader_iter to prefetch the next batch before fwd_bwd func
-        is called to avoid PP rank 2 from wait indefinitely to get outpits from PP 1
+    def _val_iterator_done(self, iterator):
+        """
+        Check if we reached trainer.limit_val_batches, if so exhaust the iterator to raise a StopIteration and exit validation_step
         """
-        elements = []
-        num_microbatches = get_num_microbatches()
-        for _ in range(num_microbatches):
+        if (
+            self._val_micro_batches_consumed == self.trainer.num_sanity_val_steps
```

Comment:
is `num_sanity_val_steps` deleted/changed after sanity check is over? 
if it is not, what if `num_sanity_val_steps = 2 `and `self.trainer.limit_val_batches = 4`. inside validation will we end after 2 batches? wouldnt that be incorrect?

Reply:
@aklife97 `_val_iterator_done` is called only after we run through all microbatches in a `validation_step`. So in case of regular validation `self._val_micro_batches_consumed` will be same as `self.trainer.num_sanity_val_steps` only if `num_sanity_val_steps == limit_val_batches`. The method never gets called in between for ex when we are done with 2 microbatches and `limit_val_batches=4`. It calls this method after finishing the remaining 2 microbatches.

Reply:
> _val_iterator_done is called only after we run through all microbatches in a validation_step

do you mean that validation_step is only run once? because we seem to be calling `_val_iterator_done` in the beginning of validation_step which would mean it runs everytime validation_step is entered?

Reply:
No `validation_step` can be run multiple times. But `self._val_micro_batches_consumed` is incremented by num of microbatches at the end of val_step and is always a multiple of num of microbatches. Let's say `limit_val_batches = 4 ` and `num_microbatches = 4` and `num_sanity_val_steps=2`. So the value of `self._val_micro_batches_consumed` will be 0 initially when `_val_iterator_done` is called, followed by `fwd_bwd_func` of val_step and then `self._val_micro_batches_consumed` is incremented to 4. So it never becomes 2 and will remain a multiple of 4  whenever it enters `_val_iterator_done` next time unless num_microbatches = 2 (and in this case it's okay if it passes the condition `self._val_micro_batches_consumed == self.trainer.num_sanity_val_steps`)

Reply:
but what if 
`num_microbatches = 2 `in the same example?

so `limit_val_batches = 4`, `num_sanity_val_steps=2 `and `num_microbatches = 2,` what will happen now?

Reply:
Yes I just realized this case, it'll break. Let me fix.

Reply:
Fixed this is the latest commit, let me know if it looks okay.

Reply:
I was observing a hang in the CI test bcoz of this issue, after the fix it runs fine.

Reply:
why do we need the if condition to check microbatches consumed if we're going to rely on iterator length?
why not just check for StopIteration and return accordingly? will also save from keeping track of _val_micro_batches_consumed

Reply:
> why do we need the if condition to check microbatches consumed if we're going to rely on iterator length? why not just check for StopIteration and return accordingly? will also save from keeping track of _val_micro_batches_consumed

We have to check if next(iter) is available num_microbatches times and since that's not a good idea (this is the prefetching logic we had previously) we want to check it just once when `_val_micro_batches_consumed` is equal to trainer.limit_val_batches or num_sanity_val_steps. Hence we are keeping track of _val_micro_batches_consumed.

Reply:
> also are we sure that PTL handles the lengths of the iterator completely correctly? if my validation set has a total of 100 batches, but limit_val_batches is set as 5 - do we have a guarantee that the iterator length will be exactly 5?
> 
> even if there is a single case where this assumption fails/PTL does not handle this properly, we can end up with huge non-trivial bugs, so would be great to double check on this assumption

Yes this has been confirmed with PTL, and they have reviewed this code too. Also, this is being added as temp snippet of code until we have PTL's patch working that'll take care of handling the end of iterator. I plan to work with PTL this week to get that patch in as soon as we can.

Reply:
the current code is equivalent to just the iterator is not empty checking code without any if condition and _val_micro_batches_consumed tracking.

we can skip the if part and the code will still work the same

Reply:
to clarify, here's what I mean: the current code is effectively just doing a stop iteration check. Removing the if-else will not change the functioning. The benefit of removing will be that all models (irrespective of downstream/pre-train) can use this function and check if the iterator is over and not worry about anything else

Reply:
I see your point, let me check on this.

Reply:
@aklife97 addressed this in the latest commit.